threefoldtech/tfgrid-sdk-ts
# Supporting deployments on zos4 nodes

### Description

Supporting deployments on zos4 nodes from grid client and dashboard
### Changes

- Added a network light and machine light workloads
- Adjusted workload types to contain the two newly added workloads
- Added unit tests for the two new workload types
- Added a ZNetworkLight primitive class to handle the creation of network light workload
- Added a VMLightPrimitive primitive class to handle the creation of zmachine light workload
- Updated the _createDeployment and create method to support zos4 deployments according to the node's features
- Updated deployMerge and saveNetworks methods in TwinDeploymentHandler
- Added getFeaturesFromFilters method in nodes and capacity module to set node's features according to the filters chosen by the user
- Added extra filters in the FilterOptions class
- Adjusted automated and manual node selection to consider features in filters
- Added planetary, mycelium and wireguard filters to all solutions
- Added a single_vm_zos4.ts script to deploy on a zos4 node
- Added deployment_flow.md 
### Related Issues

- #3462
### Tested Scenarios


- Deployed two vms with [single_vm_mycelium](https://github.com/threefoldtech/tfgrid-sdk-ts/blob/development/packages/grid_client/scripts/single_vm_mycelium.ts)  script
![Screenshot from 2024-11-17 15-26-10](https://github.com/user-attachments/assets/0ef1ccac-34b0-41a2-b259-802c6f2aa1b0)

1.  Node 168 which is a zos3 node
![Screenshot from 2024-11-17 15-27-45](https://github.com/user-attachments/assets/d3a1d676-cab0-4c23-a8eb-23a2f47306b0)

2. Node 255 which is a zos4 node
![Screenshot from 2024-11-17 15-27-20](https://github.com/user-attachments/assets/7d419a58-f19e-4267-ac45-b1e7b253f747)

- Go to Dashboard
   1. In the deploy section, select any solution
   2.  In the automated node selection:
   3. Adjust your filters to include only the Mycelium network
   4. Reload the nodes.
   5. Observe that both zos4 and zos3 nodes are listed.
   6. Deploy on any node and it will pass (note: zos4 node used is 255)
   7. Observe Deployment list
   8. Navigate to Contracts list
   
![Screenshot from 2024-11-17 15-50-10](https://github.com/user-attachments/assets/2b2ffcc7-5cb0-4bfe-9552-732c83a3ae4c)

![Screenshot from 2024-11-17 15-50-58](https://github.com/user-attachments/assets/c63f5840-0160-45b9-b231-41869c1a495a)

![Screenshot from 2024-11-17 15-52-40](https://github.com/user-attachments/assets/2c2a574f-88cb-4c56-831d-0bbdd37b47fe)

- Go to Dashboard
   1. In the deploy section, select any solution
   2. In the manual node selection:
   3. Adjust your filters to include only the Mycelium network
   4. Enter any node id 
   5. If the node has Mycelium as a feature no validation error will be shown
   6. If the node doesn't have Mycelium as  a feature  validation error will be shown


- Go to Dashboard
   1.  In the deploy section, select any solution
   2. In the automated node selection:
   3. Adjust your filters to include the Mycelium and wireguard networks
   7. Enter  node id == 259 and observe
   8. Validation error will be shown stating the the node doesn't support wireguard
 
  


### Checklist
